R/params.R

```diff
@@ -0,0 +1,329 @@
+
+knit_params_get <- function(input_lines, params) {
+  # check for recent enough knitr
+  if (packageVersion("knitr") < "1.10") {
+    stop("knitr >= 1.10 required to use rmarkdown params")
+  }
+  
+  # read the default parameters and extract them into a named list
+  knit_params <- mark_utf8(knitr::knit_params(input_lines))
+  default_params <- list()
+  for (param in knit_params) {
+    default_params[[param$name]] <- param$value
+  }
+  
+  # validate params passed to render
+  if (!is.null(params)) {
+
+    if (identical(params, "ask")) {
+      if (!interactive()) {
+        stop("render parameter configuration only allowed in an interactive environment")
+      }
+      
+      params <- knit_params_ask(input_lines = input_lines)
+      if (is.null(params)) {
+        stop("render parameter configuration canceled")
+      }
+    }
+    
+    # verify they are a list
+    if (!is.list(params) || (length(names(params)) != length(params))) {
+      stop("render params argument must be a named list")
+    }
+    
+    # verify that all parameters passed are also in the yaml
+    invalid_params <- setdiff(names(params), names(default_params))
+    if (length(invalid_params) > 0) {
+      stop("render params not declared in YAML: ",
+           paste(invalid_params, sep = ", "))
+    }
+  }
+  
+  # merge explicitly provided params with defaults
+  merge_lists(default_params, params, recursive = FALSE)
+}
+
+params_converters <- list(
+    logical = function(value) {
+      as.logical(value)
+    },
+    Date = function(value) {
+      as.Date(value)
+    },
+    POSIXct = function(value) {
+      ## FIXME: This is how knitr does datetime conversion, but maybe we should use the local tz.
+      as.POSIXct(value, tz="GMT")
+    },
+    integer = function(value) {
+      as.integer(value)
+    },
+    numeric = function(value) {
+      as.numeric(value)
+    },
+    character = function(value) {
+      as.character(value)
+    },
+    data.frame = function(value) {
+      read.csv(value$datapath)
+    }
+    )
+
+params_get_converter <- function(param) {
+  for (c in param$class) {
+    converter <- params_converters[[c]]
+    if (!is.null(converter)) {
+      return(converter)
+    }
+  }
+  ## FIXME: this probably should be an error, as it is an unsupported type.
+  return(function(value){ value })
+}
+
+params_controls <- list()
```

Comment:
The `param$class` comes directly from `param$value`. We could try dispatching on `param$value` instead of using `param$class` at all...